cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -269,30 +285,406 @@ func copyFromGlobHelper(src, dst, globEx string) error {
+        |                       |                             |                             |               |
+        |                       |success                      |success                      |               |
+        |                       |                             |                    success  |        fail   |
+        v                       v                             v                             |               |
++---+---+           +-------+----+                +-------+----+                            |               |
+|       |           | mount      |       success  | mount data |                            |               |
+| done  +<----------+ save       |      +---------+            +<---------------------------+               |
+|       |           |            |      |         |            |                                            |
++--+----+           +----+-------+      |         +----------+-+                                            |
+   ^                     ^              |                    |                                              |
+   |                     | success      v                    |                                              |
+   |                     |     +--------+----+   fail        |fail                                          |
+   |                     |     | unlock save +--------+      |                                              |
+   |                     +-----+ w/ run key  |        v      v                                              |
+   |                     ^     +-------------+   +----+------+-----+                                        |
+   |                     |                       | unlock save     |                                        |
+   |                     |                       | w/ fallback key +----------------------------------------+
+   |                     +-----------------------+                 |
+   |                             success         +-------+---------+
+   |                                                     |
+   |                                                     |
+   |                                                     |
+   +-----------------------------------------------------+
+                                                fail
+
+*/
+
+type stateMachine struct {
+	// the current state is the one that is about to be executed
+	current stateFunc
+
+	// the disk we have all our partitions on
+	disk disks.Disk
+
+	isEncryptedDev bool
+
+	// state for tracking what happens as we progress through degraded mode of
+	// recovery
+	degradedState *recoverDegradedState
+}
+
+// degraded returns whether a degraded recover mode state has fallen back from
+// the typical operation to some sort of degraded mode.
+func (m *stateMachine) degraded() bool {
+	r := m.degradedState
+	// we should have nothing in the error log
+	if len(r.ErrorLog) != 0 {
+		return true
+	}
```

Comment:
because it seems the unit tests never leave ErrorLog empty in the error cases we never tests the return true paths below, not sure it's an issue. An option would be to move this check last ??